utils/stall_free.hh

```diff
@@ -57,14 +59,176 @@ void merge_to_gently(std::list<T>& list1, const std::list<T>& list2, Compare com
+};
+
+template <typename T>
+concept Sequence = Iterable<T> && requires (T x, size_t n) {
+    { x.back() } -> std::same_as<typename T::value_type&>;
+    { x.pop_back() } -> std::same_as<void>;
+};
+
+template <typename T>
+concept TriviallyClearableSequence =
+    Sequence<T> && std::is_trivially_destructible_v<typename T::value_type> && !HasClearGentlyMethod<typename T::value_type>;
+
+template <typename T>
+concept Container = Iterable<T> && requires (T x, typename T::iterator it) {
+    { x.erase(it) } -> std::same_as<typename T::iterator>;
+};
+
+template <typename T>
+concept MapLike = Container<T> && requires (T x) {
+    std::is_same_v<typename T::value_type, std::pair<typename T::key_type, typename T::mapped_type>>;
+};
+
+template <HasClearGentlyMethod T>
+future<> clear_gently(T& o) noexcept;
+
+template <SharedPointer T>
+future<> clear_gently(T& o) noexcept;
+
+template <SmartPointer T>
+requires (!SharedPointer<T>)
+future<> clear_gently(T& o) noexcept;
+
+template <typename T, std::size_t N>
+future<> clear_gently(std::array<T, N>&a) noexcept;
+
+template <typename T>
+requires (StringLike<T> || TriviallyClearableSequence<T>)
+future<> clear_gently(T& s) noexcept;
+
+template <Sequence T>
+requires (!StringLike<T> && !TriviallyClearableSequence<T>)
+future<> clear_gently(T& v) noexcept;
+
+template <MapLike T>
+future<> clear_gently(T& c) noexcept;
+
+template <Container T>
+requires (!StringLike<T> && !Sequence<T> && !MapLike<T>)
+future<> clear_gently(T& c) noexcept;
+
+namespace internal {
+
+template <typename T>
+concept HasClearGentlyImpl = requires (T x) {
+    { clear_gently(x) } -> std::same_as<future<>>;
+};
+
+template <typename T>
+requires HasClearGentlyImpl<T>
+future<> clear_gently(T& x) noexcept {
+    return utils::clear_gently(x);
+}
+
+// This default implementation of clear_gently
+// is required to "terminate" recursive clear_gently calls
+// at trivial objects
+template <typename T>
+requires (!HasClearGentlyImpl<T>)
+future<> clear_gently(T&) noexcept {
+    return make_ready_future<>();
+}
+
+} // namespace internal
+
+template <HasClearGentlyMethod T>
+future<> clear_gently(T& o) noexcept {
+    return futurize_invoke(std::bind(&T::clear_gently, &o));
+}
+
+template <SharedPointer T>
+future<> clear_gently(T& o) noexcept {
+    if (o.use_count() == 1) {
+        return internal::clear_gently(*o);
+    }
+    return make_ready_future<>();
+}
+
+template <SmartPointer T>
+requires (!SharedPointer<T>)
+future<> clear_gently(T& o) noexcept {
+    return internal::clear_gently(*o);
+}
+
+template <typename T, std::size_t N>
+future<> clear_gently(std::array<T, N>&a) noexcept {
+    return do_for_each(a, [] (T& o) {
+        return internal::clear_gently(o);
+    });
+}
+
+// Trivially destructible elements can be safely cleared in bulk
+template <typename T>
+requires (StringLike<T> || TriviallyClearableSequence<T>)
+future<> clear_gently(T& s) noexcept {
+    s.resize(0);
```

Comment:
s.clear()?

Reply:
Clear is pointless in this case since it keeps the allocation in the vector case (and in the std::string case current implementation) and since the values are trivially destructible it achieves nothing.
Also, `seastar::sstring` doesn't implement `clear()` (it probably should to be compatible with std::string).

If the vector or string are long, we want to truncate them and delete their allocation.

Reply:
Aha.

Reply:
The documentation says:

> Vector capacity is never reduced when resizing to smaller size because that would invalidate all iterators, rather than only the ones that would be invalidated by the equivalent sequence of pop_back() calls.

Better to use `s = {};`.

Reply:
Kept in v7. Added comment explaining why we call `resize()` rather than `clear()`.

Reply:
And also add the need for clear() the TriviallyClearableSequence concept (that's the word "Clearable" in the concept's name...).

Reply:
> Better to use `s = {};`.

@avi will do

Reply:
> And also add the need for clear() the TriviallyClearableSequence concept (that's the word "Clearable" in the concept's name...).

@nyh do you mean add it to the concept requirements?
I think it should work.